pkg/util/retry/retry.go

```diff
@@ -53,6 +53,12 @@ func Do(ctx context.Context, fn func() error, opts ...Option) error {
 				return el
 			}
 
+			deadline, ok := ctx.Deadline()
+			if ok && deadline.UnixMilli()-time.Now().UnixMilli() < c.sleep.Milliseconds() {
```

Comment:
NIT: maybe following code is simpler
```suggestion
			if ok && deadline.Sub(time.Now()) < c.sleep {
```